### DISCUSSION TITLE
How to collect NetFlow data from switch?

### DISCUSSION THREAD

**Discussion by forever765** [nfstream/nfstream#63] [Q&A] [answered]
- Collector: 
OS: Centos 7.9.2009
NFStream Version: 6.3.1
Python Version: 3.6
IP Address: 192.168.123.208

- Gateway: switch
IP Address: 192.168.123.1
Netflow data will send to Collector udp/9995

Hi there,
I discovered this awesome project NFStream a few days ago, but I encountered difficulties during testing, can anyone help me? Thanks in advance!

Q1: When I try to use the example code for online mode test, after a few minutes of waiting, it returns no results.
<img width="266" alt="online_mode" src="https://user-images.githubusercontent.com/20664431/115983238-9ebf8980-a5d2-11eb-93ba-c7a4e3e6384f.png">

Q2: I have tried using tcpdump to save netflow flow data as a pcap file, But when I try to use nfstream to parse this pcap file, I can only get a flow whose src_ip is the switch ip and the dst_ip is the collector.
<img width="788" alt="pcap" src="https://user-images.githubusercontent.com/20664431/115983301-f6f68b80-a5d2-11eb-9271-8281c7f31106.png">

Q3: In the end, I tried to use tcpdump on the collector to capture all the packets and save them as pcap files. It worked, but NFStream can only be used for this purpose? Can NFStream be used as a NetFlow collector and output the results?

**Comment by drnpkr** [comment, marked as answer]
Hello,

Firstly, NFStream is not in full conformity with the IPFIX/NetFlow specification. As such, you won't be able to use it in a NetFlow fashion where you have Exporters and Collectors.

You can think of NFStream as an Exporter, which dumps the flows in its flow cache into a CSV file. You can also create a code that forwards the observed flows to a collecting process.

If you talk about NetFlow (I assume v9), you are probably familiar with active and passive (idle) timeouts. I think this is what you encounter when you do not see any flows being observed by NFStream. The default value for idle is 120s, while for passive, 180s. See the documentation here: https://www.nfstream.org/docs/api

As per your pcap experiment, did you try checking the file in Wireshark? Does it contain more than one flow? If you interrupt NFStream while processing the PCAP, the cached flows will be dumped (this is also a default behaviour). That being said, if you stop the process before it finished, your flows might be dumped.

**Reply by forever765** [reply]
Hi @drnpkr ,
I am looking for a NetFlow/sFlow collector that supports nDPI, NFStream does not seem to be a flow collector, which is very regrettable.
But anyway, thank you for your reply!

**Reply by aouinizied** [reply]
ntopng can be an interesting option for you. We are planning to extend nfstream with ntopng export interface.

**Reply by forever765** [reply]
Thank you for your advice. Unfortunately, We are not interested in paid software. So we are using nfdump as the flow collector. According to your suggestions, I will learn how to use nDPI, Thanks!
